django/contrib/auth/__init__.py

```diff
@@ -199,8 +199,11 @@ def get_user(request):
             # Verify the session
             if hasattr(user, "get_session_auth_hash"):
                 session_hash = request.session.get(HASH_SESSION_KEY)
-                session_hash_verified = session_hash and constant_time_compare(
-                    session_hash, user.get_session_auth_hash()
+                session_hash_verified = session_hash and any(
+                    constant_time_compare(
+                        session_hash, user.get_session_auth_hash(secret)
+                    )
+                    for secret in [settings.SECRET_KEY, *settings.SECRET_KEY_FALLBACKS]
```

Comment:
I do not think that this introduces a timing attack. Sure, you will know that the session does not match, but in this case there will most likely be further codepaths down the road that make that difference explicit. The idea of `constant_time_compare` is to protect against brute-forcing the session id byte by byte, the new code still protects against that. What do you think?

Btw I assume `any` and the generator expression short-circuit, yes? This certainly deserves a comment to explain why this only runs till it finds a match…

Reply:
> Btw I assume any and the generator expression short-circuit, yes?

Yes precisely.
The logic changed slightly, it should now be more "obvious" when reading that the loop is only called in case the first secret doesn't match, and that it breaks out once it found one working secret.

> not reintroduce a possible timing attack

It's a good point. I think that one would be able to detect which fallback secret position/number matches the given session hash.
But I don't think it is sensitive information. It informs an attacker perhaps how many/how often secret are rotated, but I'm not sure this could be leveraged directly to compromise anything 🤔